### ISSUE TITLE
android_binary cannot build if java8 Collector.of is referenced

### ISSUE THREAD

**Issue by bdleitner** [bazelbuild/bazel#12944]
### Description of the problem / feature request:

Attempting to `bazel build` an `android_binary` target that references Java8's `Collector.of` method fails with an error message saying:

```
invoking a static interface method java.util.stream.Collector.of:(Ljava/util/function/Supplier;Ljava/util/function/BiConsumer;Ljava/util/function/BinaryOperator;Ljava/util/
function/Function;[Ljava/util/stream/Collector$Characteristics;)Ljava/util/stream/Collector; strictly requires --min-sdk-version >= 24 (blocked at current API level 13)
```

Even when all `AndroidManifest.xml` files specify `minSdkVersion`s at least 24.  So at best this is a misleading error message.

I am not 100% certain the problem is with `bazel` as opposed to something in `com.google.devtools.build.android`, but since `Collector.of` is explicitly given as supported in the [Java 8 support table](https://developer.android.com/studio/write/java8-support-table), this is as deep as I've been able to reach in trying to pinpoint the problem.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Minimal example: https://github.com/bdleitner/android_error_demo

### What operating system are you running Bazel on?

Windows 10 Pro Build 10.0.19041

### What's the output of `bazel info release`?

release 4.0.0

###  Have you found anything relevant by searching the web?

No.  I started with searching, then posted a question to [StackOverflow](https://stackoverflow.com/questions/65875620/bazel-build-android-binary-with-guava-dependency-complains-about-min-sdk-version) (before I had pinned down the cause as well as I have now) which was quickly marked duplicate to something unrelated.  After more investigation I followed up with a [modified StackOverflow question](https://stackoverflow.com/questions/65947956/bazel-android-java8-complains-about-non-referenced-api-level-13), which also originally referred to `Guava` but has since been modified to reflect the minimal example.

----
### Any other information, logs, or outputs that you want to share?

This is likely an unrelated issue, but I also tried running with `--nodesugar_for_android` to see if I could just build with Java8.  Sometimes I get warnings that java8 features are not supported on `-source 7` and this persists even with the `javacopt` command line flag and with `javacopts` added to the build targets.  `Bazel` ignoring `-source 8` seems to be tied to something in the `WORKSPACE` as in my minimal example this stopped happening once I got rid of everything in the `WORKSPACE` file except for `android_sdk_repository`.  I haven't had a chance yet to isolate what specifically in the `WORKSPACE` was causing `bazel` to ignore the `javacopt` flags.  But in the minimal example where this does not occur I get a different error:

```
 Building src/main/java/com/example/bazel/libgreeter_activity.jar (5 source files) failed: (Exit 1): java.exe failed: error executing command external/remotejdk11_win/bin/java.exe -XX:+UseParalle
lOldGC -XX:-CompactStrings --patch-module=java.compiler=external/remote_java_tools_windows/java_tools/java_compiler.jar ... (remaining 15 argument(s) skipped)
src\main\java\com\example\bazel\MainActivity.java:26: error: cannot find symbol
    clickMeButton.setOnClickListener(v -> helloBazelTextView.setText(greeter.sayHello()));
                                     ^
  symbol:   method metafactory(Lookup,String,MethodType,MethodType,MethodHandle,MethodType)
  location: interface LambdaMetafactory
Fatal Error: Unable to find method metafactory
```

Searching for that led me to [this StackOverflow question](https://stackoverflow.com/questions/48427655/cannot-find-symbol-method-metafactory-using-lambda), which references `core-lambda-stubs.jar`. I verified that I do have this file (dated 2021-01-18) in my latest build_tools (30.0.3) folder.  Specifying `build_tools_version = "30.0.3"` in my `WORKSPACE` (just in case) did not help.

**Comment by bdleitner** [comment]
I had implemented a workaround to use the guava_android target and had my own custom methods to handle collecting to an immutable set or list (that didn't use `Collector.of`

But then in another, larger project I ran into a similar problem with `Comparator.comparing`.  This seems quite odd as it's also just another static interface method, when other such methods can be used without a problem. 

I'm now wondering if there's an issue where anything in the java core libraries is assumed to have a min-sdk-version of 13.  I have updated my [minimal example](https://github.com/bdleitner/android_error_demo) to include a new class that demonstrates the break.
